### ISSUE TITLE
Remove phly/http coupling

### ISSUE THREAD

**Issue by sagikazarmark** [php-http/adapter-core#5]
Currently we rely on phly/http. The custom message extensions are removed, but we still have a few things:

- Internal request
- Message factory

The internal request could be solved by one of the following ways:

Is it really necessary to implement PsrRequestInterface in InternalRequest? If yes, then we have to either rely on one implementation or create one for all common packages and require one of them. If no, then we can remove the coupling there. (My guess is that there is a reason why internal request is PSR request as well.

An alternative solution is to make the internal request a decorator as well. However in this case we still have to know the exact implementation installed.

The other thing is the message factory: Message factory is hardly coupled to the underlying classes, however there is a way to create a generic factory: use Reflection and create the class without calling the constructor and pass everything through the interface methods. However we also have to know the exact implementation to do this.

One possible solution is that during composer package installation loop through all classes and find PSR implementations and store that information somewhere. But rather not do that.

To sum up, it is not easy at all to remove coupling to a specific implementation.

**Comment by sagikazarmark** [comment]
Here is what we can do:

- Make the internal request a decorator
- Implement a few "known" Message Factories in the core package
- If no message factories are passed to the constructor, the `MessageFactoryGuesser` checks if one of the implemented message factories can be used (aka the compatible package is installed). If not, an exception is thrown

**Comment by ddeboer** [comment]
So will php-data include its own implementation of the PSR-7 RequestInterface? Or do you prefer to use depend on another library (such as [diactoros](https://github.com/zendframework/zend-diactoros)) that provides Request and Response objects?

**Comment by sagikazarmark** [comment]
I guess you mean php-http. ;)

Currently we rely on phly/http.

The plan is to support any package that provides `psr/http-message-implementation`. However this is a bit hard, because we also need a message factory which MUST be aware of the supported objects (at least their name). So I have multiple solutions in my mind:

- Allow any message implementations to be installed, but require the user to implement (and pass) the message factory. This can be combined with the solution in my previous comment (MessageFactoryGuesser for some message factories contained in the core out of the box)
- Create a custom virtual package that provides our own virtual package: php-http/message-implementation. This would mean that for every PSR7 implementations out there we have to create an adapter package. While this might be easier from a point of dependency, it absolutely adds complexity which I would like to avoid.

BTW thanks for pointing us to diactoros, good to know phly/http is (going to be) deprecated.
